### DISCUSSION TITLE
Parsing Union of objects removes property from object

### DISCUSSION THREAD

**Discussion by WayneEllery** [colinhacks/zod#2763] [Q&A] [answered]
```TypeScript
import { z } from "zod";

const UpdateBookingSchema = z.union([
  z.object({
    propertyType: z.string()
  }),
  z.object({
    region: z.string()
  }),
  z.object({
    propertyType: z.string(),
    region: z.string()
  })
]);

const updateBooking = UpdateBookingSchema.parse({
  propertyType: "unit",
  region: "vic_bayside"
});

console.log({ updateBooking });
```

Output:
```
{ propertyType: "unit" }
```

Expected:
```
{
  propertyType: "unit",
  region: "vic_bayside"
}
```

https://codesandbox.io/s/hungry-mcclintock-9dzmnp

If I move the last `z.object` to the first item in the union it works as expected

**Comment by colinhacks** [comment]
This is behaving as expected. Zod iterates over the union schemas until it finds one that parses successfully. If multiple schemas pass validation, Zod has no way of picking which schema is "more correct" in the general case.

**Comment by WayneEllery** [comment]
Thanks Colin. That makes sense. Could it first try the objects with the most properties?

**Comment by stevegcarpenter** [comment]
@WayneEllery , try adding a `.strict()` on the end of each object in the union. 
https://codesandbox.io/s/polished-silence-fz8lpr

**Comment by WayneEllery** [comment]
> @WayneEllery , try adding a `.strict()` on the end of each object in the union. https://codesandbox.io/s/polished-silence-fz8lpr

That does work but that also prevents additional data being in the objects which is not desired.

**Comment by JacobWeisenburger** [comment, marked as answer]
@WayneEllery

Will this work for you?

```ts
const UpdateBookingSchema = z.union( [
    z.object( {
        propertyType: z.string(),
        region: z.string()
    } ),
    z.object( {
        propertyType: z.string()
    } ),
    z.object( {
        region: z.string()
    } ),
] )

console.log(
    UpdateBookingSchema.parse( {
        propertyType: 'unit',
        region: 'vic_bayside'
    } )
)
// {
//     propertyType: "unit",
//     region: "vic_bayside"
// }
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
